### PR TITLE
Start dirty page tracking implementation in vm-memory.

### PR DESCRIPTION
**Pull request by sandreim**
## Reason for This PR

Bootstrap implementation for dirty page tracking in vm-memory - #1997

Thanks @alexandruag for providing insight/PoC on the proxy crate approach.

## Description of Changes

Create a "proxy" crate that Firecracker will use as a dependency for all current vm-memory imports. This crate will export local GuestMemoryMmap and GuestRegionMap implementations that will contain dirty page tracking functionality. All other re-exports will be fetched from upstream vm-memory.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/arch/Cargo.toml [hunk 1]
```diff
@@ -8,8 +8,7 @@ edition = "2018"
 kvm-bindings = { git = "https://github.com/firecracker-microvm/kvm-bindings", tag = "v0.2.0-2", features = ["fam-wrappers"] }
 kvm-ioctls = { git = "https://github.com/firecracker-microvm/kvm-ioctls", tag = "v0.5.0-2" }
 libc = ">=0.2.39"
-vm-memory = { version = ">=0.2.2", features = ["backend-mmap"] }
-
+vm-memory-proxy = { path = "../vm-memory-proxy" }
```
Comment (serban300) [inline src/arch/Cargo.toml:11]: Can we keep the old name (`vm-memory`) ? Chaning it has 2 disadvantages:
1. It's very verbose
2. We'll have to undo this change when we switch back to the `vm-memory` crate.

Comment (sandreim) [reply src/arch/Cargo.toml:11]: I agree on both points. My idea here was to make the switch be explicit and much clear.

**Comment by aghecenco** [comment]
I may have gotten it backwards - I was thinking that Firecracker will *continue* consuming `rust-vmm/vm-memory`, but instead of using `rust-vmm/vm-memory::GuestMemoryMmap` all over the place, Firecracker would add its own `struct GuestMemoryMmapWithTracking` (or something) and `impl rust-vmm/vm-memory::GuestMemory for firecracker::GuestMemoryMmapWithTracking` (disregard rustlang blasphemy). That way we don't have to kill the dependency. Did we move away from that? What's the advantage of removing the dependency, other than renaming `GuestMemoryMmap` everywhere?

**Comment by sandreim** [comment]
> I may have gotten it backwards - I was thinking that Firecracker will _continue_ consuming `rust-vmm/vm-memory`, but instead of using `rust-vmm/vm-memory::GuestMemoryMmap` all over the place, Firecracker would add its own `struct GuestMemoryMmapWithTracking` (or something) and `impl rust-vmm/vm-memory::GuestMemory for firecracker::GuestMemoryMmapWithTracking` (disregard rustlang blasphemy). That way we don't have to kill the dependency. Did we move away from that? What's the advantage of removing the dependency, other than renaming `GuestMemoryMmap` everywhere?

The dependency to vm-memory from upstream is kept in the local proxy crate `vm-memory-proxy` and we override only what is needed with a local implementation. The rest of the FC code depends on the proxy crate which re-exports the upstream stuff.

File: src/arch/Cargo.toml (hunk 1 above)
Comment (sandreim) [reply src/arch/Cargo.toml]: I've updated the PR and now the proxy crate has the same name: `vm-memory`

Comment (acatangiu) [reply src/arch/Cargo.toml]: Only 11 files changed now as opposed to the previous version with 50 files changed 👍

**Comment by aghecenco** [comment]
> we override only what is needed with a local implementation

Why not create a new implementor for `GuestMemory`?

**Comment by raduweiss** [comment]
> > we override only what is needed with a local implementation
> 
> Why not create a new implementor for `GuestMemory`?

FWIW I also understood the coupling in the way @aghecenco described it above.

**Comment by sandreim** [comment]
> > > we override only what is needed with a local implementation
> > 
> > 
> > Why not create a new implementor for `GuestMemory`?
> 
> FWIW I also understood the coupling in the way @aghecenco described it above.

The goal here is to do as few changes as possible to the interfaces we currently use in Firecracker with the hope that tracking will be available in upstream. The Firecracker code uses both the GuestMemory  interface and also the GuestMemoryMmap directly. If we implement a new backend like GuestMemoryMmapTracked we would need to do more changes in Firecracker.

The approach in this PR is less invasive and completely transparent to Firecracker as we re-export all the stuff from upstream and we override only what's needed to support our usecase.

File: src/vm-memory/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "vm-memory"
+version = "0.1.0"
+authors = ["Amazon Firecracker team <firecracker-devel@amazon.com>"]
+edition = "2018"
+
+[dependencies]
+vm-memory = { version = ">=0.2.2", features = ["backend-mmap"] }
```
Comment (aghecenco) [inline src/vm-memory/Cargo.toml:8]: This
```suggestion
vm-memory-upstream = { package = "vm-memory", version = ">=0.2.2", features = ["backend-mmap"] }
```
and replacing occurrences of `vm_memory` with `vm_memory_upstream` inside the `firecracker/vm_memory` crate should do the trick for the doctests

Comment (sandreim) [reply src/vm-memory/Cargo.toml]: Thanks, this fixes the doctest issue.

File: src/vm-memory/src/bitmap.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    /// Get the length of the bitmap in bits (i.e. in how many pages it can represent).
+    pub fn len(&self) -> usize {
+        self.size
+    }
+
+    /// Is the bitmap empty (i.e. has zero size)? This is always false, because we explicitly
+    /// round up the size when creating the bitmap.
+    pub fn is_empty(&self) -> bool {
+        false
+    }
```
Comment (luminitavoicu) [inline src/vm-memory/src/bitmap.rs:79]: I don't think this function is needed, since we know the bitmap is never empty. WDYT?

Comment (sandreim) [reply src/vm-memory/src/bitmap.rs:79]: Makes sense, I'll remove it.

**Review by luminitavoicu — APPROVED** [review]

File: src/vm-memory/src/mmap.rs [hunk 4]
```diff
@@ -0,0 +1,1184 @@
...
+        .unwrap();
+        assert!(gm.find_region(start_addr).is_some());
+        let region = gm.find_region(start_addr).unwrap();
+        assert!(region.file_offset().is_some());
+    }
+
+    // Windows needs a dedicated test where it will retrieve the allocation
+    // granularity to determine a proper offset (other than 0) that can be
+    // used for the backing file. Refer to Microsoft docs here:
+    // https://docs.microsoft.com/en-us/windows/desktop/api/memoryapi/nf-memoryapi-mapviewoffile
```
Comment (luminitavoicu) [inline src/vm-memory/src/mmap.rs:1098]: I don't think we should we keep this indication.

Comment (sandreim) [reply src/vm-memory/src/mmap.rs:1098]: You are right.

File: src/vm-memory/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
```
Comment (acatangiu) [inline src/vm-memory/src/lib.rs:6]: nit: is this still the case?

File: src/vm-memory/src/mmap.rs [hunk 6]
```diff
@@ -0,0 +1,1191 @@
+// Copyright (C) 2019 Alibaba Cloud Computing. All rights reserved.
+//
+// Portions Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (acatangiu) [inline src/vm-memory/src/mmap.rs:3]: nit: For this local file where we add dirty tracking support, these two might be better reversed.

File: src/vm-memory/src/bitmap.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+impl Bitmap {
+    /// Create a new bitmap of `byte_size`, with one bit per `page_size`.
+    /// In reality this is rounded up, and you get a new vector of the next multiple of 64 bigger
+    /// than `size` for free.
+    pub fn new(byte_size: usize, page_size: usize) -> Self {
+        // Bit size is the number of bits in the bitmap, always at least 1 (to store the state of
+        // the '0' address).
+        let bit_size = std::cmp::max(1, byte_size / page_size);
```
Comment (acatangiu) [inline src/vm-memory/src/bitmap.rs:28]: nit:
```suggestion
        let num_pages = std::cmp::max(1, byte_size / page_size);
```
for clarity.

File: src/vm-memory/src/mmap.rs [hunk 8]
```diff
@@ -206,14 +200,27 @@ impl GuestMemoryRegion for GuestRegionMmap {
...
+        // This is safe because we mapped the area at addr ourselves, so this slice will not
+        // overflow. However, it is possible to alias.
+        Some(std::slice::from_raw_parts(
+            self.mapping.as_ptr(),
+            self.mapping.size(),
+        ))
     }
 
+    // TODO: This implementation is temporary.
+    // We need to return None here once we refactor vsock.
```
Comment (acatangiu) [inline src/vm-memory/src/mmap.rs:215]: Let's add an issue to track this.

**Review by acatangiu — APPROVED** [review]

File: src/vm-memory/src/mmap.rs (hunk 8 above)
Comment (sandreim) [reply src/vm-memory/src/mmap.rs:215]: Created #2153

**Pull request merged**